### PR TITLE
Bug 1888525: Adding capacity to the storage cluster [RHOCS Usability Study]

### PR DESCRIPTION
**Pull request by SanjalKatiyar**
1) Added alert-action for adding capacity on the status-card.
Added for two alert-types: `CephClusterNearFull` & `CephClusterCriticallyFull`;
2) Made changes to raw-capacity card, donut-chart & chart-legend will change color based on available capacity.
Danger threshold is when storage-cluster is > 80% utilised and,
Warning threshold is when storage-cluster utilisation is > 75% && <= 80%;

UI will look something like this after the changes:
Screenshot from 2021-04-09 18-19-01
Screenshot from 2021-04-09 18-16-11

### CODE REVIEW

File: frontend/packages/ceph-storage-plugin/console-extensions.json [hunk 1]
```diff
@@ -20,5 +20,27 @@
     "flags": {
       "required": ["LSO_DEVICE_DISCOVERY", "OCS_ATTACHED_DEVICES"]
     }
+  },
+  {
+    "type": "console.alert-action",
+    "properties": {
+      "alert": "CephClusterNearFull",
+      "text": "%ceph-storage-plugin~Add Capacity%",
+      "action": { "$codeRef": "alert.getAlertActionFunction" }
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/console-extensions.json:29]:
```suggestion
      "action": { "$codeRef": "alert.launchClusterExpansionModal" }
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 2]
```diff
@@ -10,11 +10,34 @@ import DashboardCardBody from '@console/shared/src/components/dashboard/dashboar
...
 import './raw-capacity-card.scss';
 
 const queries = (() => Object.values(CAPACITY_INFO_QUERIES))();
 const colorScale = ['#0166cc', '#d6d6d6'];
+let capacityStatus = '';
+
+const setCapacityStatus = (capacityRatio) => {
+  if (capacityRatio > DANGER_THRESHOLD) {
+    capacityStatus = DANGER;
+    colorScale[0] = '#C9190B';
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:33]: nit: make hex code lowercase

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/common/capacity-breakdown/consts.ts [hunk 3]
```diff
@@ -15,3 +15,7 @@ export const COLORMAP = [
 export const OTHER = 'Other';
 export const CLUSTERWIDE = 'Cluster-wide';
 export const BUCKETCLASSKIND = 'BucketClass';
+export const DANGER = 'Danger';
+export const WARNING = 'Warning';
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/common/capacity-breakdown/consts.ts:19]:
```suggestion
export const DANGER = 'Danger';
export const WARNING = 'Warning';
```
These need to be translated as well.

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 4]
```diff
@@ -10,11 +10,34 @@ import DashboardCardBody from '@console/shared/src/components/dashboard/dashboar
...
+  } else if (capacityRatio > WARNING_THRESHOLD && capacityRatio <= DANGER_THRESHOLD) {
+    capacityStatus = WARNING;
+    colorScale[0] = '#F0AB00';
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:36]: nit: hex code should be lower case

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 5]
```diff
@@ -10,11 +10,34 @@ import DashboardCardBody from '@console/shared/src/components/dashboard/dashboar
...
+  } else {
+    capacityStatus = '';
+  }
+};
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:40]: move them inside of the component. Make color scale a React state. Set capacity state should run inside a React useEffect. Should run every time `capacityRatio` changes.

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/common/capacity-breakdown/consts.ts (hunk 3 above)
Comment (SanjalKatiyar) [reply frontend/packages/ceph-storage-plugin/src/components/dashboards/common/capacity-breakdown/consts.ts:19]: these are constants... I am not displaying them anywhere, I am using it for checking ternary conditions...
By "translated" do you mean: `export const DANGER = 'danger'`??

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 6]
```diff
@@ -10,11 +10,15 @@ import DashboardCardBody from '@console/shared/src/components/dashboard/dashboar
...
+import {
+  RedExclamationCircleIcon,
+  YellowExclamationTriangleIcon,
+} from '@console/shared/src/components/status/icons';
 import { CAPACITY_INFO_QUERIES } from '../../../../queries';
+import { DANGER_THRESHOLD, WARNING_THRESHOLD } from '../../common/capacity-breakdown/consts';
 import './raw-capacity-card.scss';
 
 const queries = (() => Object.values(CAPACITY_INFO_QUERIES))();
-const colorScale = ['#0166cc', '#d6d6d6'];
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:22]:
```suggestion
const generalColorScale = ['#0166cc', '#d6d6d6'];
const warningColorScale = ['#f0ab00', '#d6d6d6'];
const dangerColorScale = ['#c9190b', '#d6d6d6'];
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 7]
```diff
@@ -24,6 +28,9 @@ const RawCapacityCard: React.FC = React.memo(() => {
   const loadError = values.every((item) => typeof item === 'undefined' || item === 0);
   const { t } = useTranslation();
 
+  const [colorScale, setColorScale] = React.useState<string[]>(['#0166cc', '#d6d6d6']);
+  const [capacityStatus, setCapacityStatus] = React.useState<JSX.Element>(null);
+
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:33]:
```suggestion
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 8]
```diff
@@ -39,6 +46,21 @@ const RawCapacityCard: React.FC = React.memo(() => {
...
+      setColorScale((arr) => ['#c9190b', arr[1]]);
+    } else if (capacityRatio > WARNING_THRESHOLD && capacityRatio <= DANGER_THRESHOLD) {
+      setCapacityStatus(<YellowExclamationTriangleIcon title="Warning" />);
+      setColorScale((arr) => ['#f0ab00', arr[1]]);
+    } else {
+      setCapacityStatus(null);
+      setColorScale((arr) => ['#0166cc', arr[1]]);
+    }
+  }, [capacityRatio, t]);
+
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:63]:
```suggestion

  const colorScale = React.useMemo(() => {
    if (capacityRatio > DANGER_THRESHOLD) {
      return dangerColorScale;
    } else if (capacityRatio > WARNING_THRESHOLD && capacityRatio <= DANGER_THRESHOLD) {
      return warningColorScale;
    } else {
      return generalColorScale;
    }
  }, [capacityRatio]);
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 9]
```diff
@@ -39,6 +46,21 @@ const RawCapacityCard: React.FC = React.memo(() => {
   const usedCapacityAdjusted = humanizeBinaryBytes(usedCapacityMetric);
   const availableCapacityAdjusted = humanizeBinaryBytes(totalCapacityMetric - usedCapacityMetric);
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:48]:
```suggestion
const CapacityStatusIcon: React.FC<CapacityStatusIconProps> = ({ratio}) => {
const { t } = useTranslation();
if (ratio < WARNING_THRESHOLD) return null;
return (
{capacityRatio > DANGER_THRESHOLD && <RedExclamationCircleIcon title={t("ceph-storage-plugin~Error")} />}
{(capacityRatio > WARNING_THRESHOLD && capacityRatio <= DANGER_THRESHOLD) &&  <YellowExclamationTriangleIcon title={t("ceph-storage-plugin~Warning")} />}
)
}
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 10]
```diff
@@ -74,6 +96,7 @@ const RawCapacityCard: React.FC = React.memo(() => {
                 fill={colorScale[1]}
                 title={t('ceph-storage-plugin~Available')}
                 text={availableCapacityAdjusted.string}
+                capacityStatus={capacityStatus}
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:99]:
```suggestion
                capacityStatus={<CapacityStatusIcon ratio={capacityRatio} />}
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 11]
```diff
@@ -39,6 +46,21 @@ const RawCapacityCard: React.FC = React.memo(() => {
   const usedCapacityAdjusted = humanizeBinaryBytes(usedCapacityMetric);
   const availableCapacityAdjusted = humanizeBinaryBytes(totalCapacityMetric - usedCapacityMetric);
 
+  const capacityRatio = usedCapacity.value / totalCapacity.value;
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:49]:
```suggestion
  const capacityRatio = parseFloat((usedCapacity.value / totalCapacity.value).toFixed(2));
```
Should avoid re-renders for very small changes in storage usage.

**Review by bipuladh — CHANGES_REQUESTED** [review]
Please run `yarn i18n`

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 12]
```diff
@@ -10,20 +10,43 @@ import DashboardCardBody from '@console/shared/src/components/dashboard/dashboar
...
+};
+
 const RawCapacityCard: React.FC = React.memo(() => {
   const [values, loading] = usePrometheusQueries(queries, parser as any);
   const loadError = values.every((item) => typeof item === 'undefined' || item === 0);
   const { t } = useTranslation();
 
+  const generalColorScale = ['#0166cc', '#d6d6d6'];
+  const warningColorScale = ['#f0ab00', '#d6d6d6'];
+  const dangerColorScale = ['#c9190b', '#d6d6d6'];
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:48]: Keep it outside of the component. If you do that, you won't have to add it to the dependency list of useMemo.

File: frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx [hunk 13]
```diff
@@ -10,20 +10,43 @@ import DashboardCardBody from '@console/shared/src/components/dashboard/dashboar
...
+import { DANGER_THRESHOLD, WARNING_THRESHOLD } from '../../common/capacity-breakdown/consts';
 import './raw-capacity-card.scss';
 
 const queries = (() => Object.values(CAPACITY_INFO_QUERIES))();
-const colorScale = ['#0166cc', '#d6d6d6'];
 
 // Enchance instantVectorStats to directly parse the values (else loading state won't be accurate)
 const parser = compose((val) => val?.[0]?.y, getInstantVectorStats);
 
+const CapacityStatusIcon: React.FC<CapacityStatusIconProps> = ({ ratio }) => {
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/raw-capacity-card/raw-capacity-card.tsx:26]:
```suggestion
const CapacityStatusIcon: React.FC<CapacityStatusIconProps> = React.memo({ ratio }) => {
```

**Review by bipuladh — CHANGES_REQUESTED** [review]

**Review by bipuladh — APPROVED** [review]
/lgtm

**Pull request merged**